packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx

```diff
@@ -0,0 +1,214 @@
+  FieldsetHTMLAttributes,
+  InputHTMLAttributes,
+  Ref,
+  forwardRef,
+} from 'react';
+
+import styled from '../../styles/styled';
+import { typography } from '../../styles/style-mixins';
+import { uniqueId } from '../../util/id';
+import { Checkbox, CheckboxProps } from '../Checkbox/Checkbox';
+import {
+  FieldWrapper,
+  FieldLabelText,
+  FieldValidationHint,
+} from '../FieldAtoms';
+import { AccessibilityError } from '../../util/errors';
+
+export interface CheckboxGroupProps
+  extends Omit<FieldsetHTMLAttributes<HTMLFieldSetElement>, 'onChange'> {
+  /**
+   * A collection of available options. Each option must have at least a label and a value
+   * for the respective Checkbox.
+   */
+  options: (Omit<CheckboxProps, 'onChange'> & {
+    label: string;
+  })[];
+  /**
+   * The values of the Checkboxes that are checked by default.
+   */
+  value?: CheckboxProps['value'][];
+  /**
+   * A callback that is called when any of the checkboxes change their values.
+   * Passed on to the Checkboxes.
+   */
+  onChange: CheckboxProps['onChange'];
+  /**
+   * A description of the selector group.
+   */
+  label: string;
+  /**
+   * Label to indicate that the input group is optional. Only displayed when the
+   * `required` prop is falsy.
+   */
+  optionalLabel?: string;
+  /**
+   * The ref to the HTML DOM element.
+   */
+  ref?: Ref<HTMLFieldSetElement>;
+  /**
+   * Makes the input group required.
+   */
+  required?: InputHTMLAttributes<HTMLInputElement>['required'];
+  /**
+   * An information, warning or error message, displayed below the Checkboxes.
+   */
+  validationHint?: string;
+  /**
+   * Triggers error message below the Checkboxes.
+   */
+  invalid?: boolean;
+  /**
+   * Triggers warning message below the Checkboxes.
+   */
+  hasWarning?: boolean;
+  /**
+   * Triggers valid message below the Checkboxes.
+   */
+  showValid?: boolean;
+  /**
+   * Visually hide the label. This should only be used in rare cases and only
+   * if the purpose of the field can be inferred from other context.
+   */
+  hideLabel?: boolean;
+}
+
+const Legend = styled('legend')<Record<string, unknown>>(typography('two'));
+
+/**
+ * A group of Checkboxes.
+ */
+export const CheckboxGroup = forwardRef(
+  (
+    {
+      options,
+      value,
+      onChange,
+      'name': customName,
+      label,
+      invalid,
+      validationHint,
+      showValid,
+      disabled,
+      hasWarning,
+      hideLabel,
+      optionalLabel,
+      required,
+      'aria-describedby': descriptionId,
+      ...props
+    }: CheckboxGroupProps,
+    ref: CheckboxGroupProps['ref'],
+  ) => {
+    if (
+      process.env.NODE_ENV !== 'production' &&
+      process.env.NODE_ENV !== 'test' &&
+      !label
+    ) {
+      throw new AccessibilityError(
+        'CheckboxGroup',
+        'The `label` prop is missing. Pass `hideLabel` if you intend to hide the label visually.',
+      );
+    }
+    const name = customName || uniqueId('checkbox-button-group_');
```

Comment:
Thanks for spotting this! We can have a "custom" `name` (i.e., `checkbox-button-group_`${name}``) for the whole checkbox group and allow developers to pass the names for the individual checkboxes via the `options` prop. I don't think it's correct to have all the checkboxes in the checkbox group having the same name. I took inspirations from the implementation of the `RadioButtonGroup` where the individual buttons should share the same `name` as the group itself (which has a custom name) for the radio button group to work correctly. This is obviously not the case for the CheckboxGroup.
I've stopped passing the checkbox group's customName to the individual checkboxes. Let me know if you think this could be implemented better!